rpcclient/roundrobin.go

```diff
@@ -0,0 +1,434 @@
+	if rrb.IsAllDisconnected() {
+		rrb.NeedsClientRestart = true
+		return true
+	}
+	return false
+}
+
+// CloseAll closes all the socket connections in the wsConns list
+// This must be called during shutdown.
+func (rrb *RoundRobinBalancer) CloseAll() {
+	rrb.mu.Lock()
+	for host := range rrb.wsConns {
+		if rrb.connState[host] != disconnected && rrb.connState[host] != shutdown {
+			log.Tracef("Balancer: Disconnecting RPC client %s", host)
+			rrb.NotifyConnStateChange(rrb.hostAddMap[host], disconnected, false)
+			wsConn, connOk := rrb.wsConns[host]
+			if connOk {
+				err := wsConn.Close()
+				if err != nil {
+					log.Errorf("Failed disconnecting to %s: %v", host, err)
+				}
+			}
+
+		}
+	}
+	rrb.mu.Unlock()
+}
+
+// NotifyReconnect will update the map for wsConns and update the connection state
+// for the corresponding host address.
+func (rrb *RoundRobinBalancer) NotifyReconnect(wsConn *websocket.Conn, hostAdd *HostAddress) {
+	rrb.mu.Lock()
+	rrb.connState[hostAdd.Host] = ready
+	rrb.wsConns[hostAdd.Host] = wsConn
+	rrb.hostAddMap[hostAdd.Host].retryCount = 0
+	rrb.mu.Unlock()
+	rrb.WsInHandler(wsConn, hostAdd.Host)
+}
+
+// UpdateReconnectAttempt will increase retryattempt + 1
+// for corresponding host address.
+func (rrb *RoundRobinBalancer) UpdateReconnectAttempt(hostAdd *HostAddress) {
+	rrb.mu.Lock()
+	rrb.hostAddMap[hostAdd.Host].retryCount++
+	rrb.mu.Unlock()
+}
+
+// IsAllDisconnected would return true if all hostaddresses are marked with
+// connection state as Disconnected.
+func (rrb *RoundRobinBalancer) IsAllDisconnected() bool {
+	rrb.mu.Lock()
+	for _, state := range rrb.connState {
+		if state != disconnected && state != shutdown {
+			rrb.mu.Unlock()
+			return false
+		}
+	}
+	rrb.mu.Unlock()
+	return true
+}
+
+// NextDisconnectedWsConn will iterate over the connection state map and
+// return the first wsConnection that has its state as Disconnected.
+func (rrb *RoundRobinBalancer) NextDisconnectedWsConn() *HostAddress {
+	rrb.mu.Lock()
+	for host, state := range rrb.connState {
+		if state == disconnected {
+			rrb.mu.Unlock()
+			return rrb.hostAddMap[host]
+		}
+	}
+	rrb.mu.Unlock()
+	return nil
+}
+
+// ConnectionState will return the connection state for given host.
+func (rrb *RoundRobinBalancer) ConnectionState(host string) (state ConnectionState, ok bool) {
+	rrb.mu.Lock()
+	state, ok = rrb.connState[host]
+	rrb.mu.Unlock()
+	return
+}
+
+// WsConnection gets the wsConn corresponding to the host.
+func (rrb *RoundRobinBalancer) WsConnection(host string) (wsConn *websocket.Conn, connOk bool) {
+	rrb.mu.Lock()
+	wsConn, connOk = rrb.wsConns[host]
+	rrb.mu.Unlock()
+	return
+}
+
+// BuildBalancer is used to setup balancer with required config.
+// If no items in HostAddresses list, then would add one using *ConnConfig.Host
+// and *ConnConfig.Endpoint else use the *ConnConfig.HostAddresses for round-robin.
+func (cn *Client) BuildBalancer(cc *ConnConfig) *RoundRobinBalancer {
```

Comment:
Please don't use variable names that are inconsistent with the rest of the package. The receiver for the *Client is `c` in every other instance in the package.  Also, the parameter name for *ConnConfig is `config` in every other instance in the package.

Reply:
This, have resolved in my latest changes.